File: copilot-connector/sample_codes/typescript/common-patterns/incremental-sync.ts
// Incremental sync pattern: track changes and only re-ingest modified items
// Shows: timestamp-based incremental sync with checkpoint tracking,
//   hash-based change detection using Node.js crypto module
//
// Prerequisites: npm install @microsoft/microsoft-graph-client @azure/identity

import { createHash } from "node:crypto";
import { Client } from "@microsoft/microsoft-graph-client";

/** A source item from the external data store. */
interface SourceItem {
  id: string;
  title: string;
  status: string;
  fullContent: string;
  url: string;
  iconUrl?: string;
  lastModified: Date;
}

/** Checkpoint persisted between sync runs. */
interface SyncCheckpoint {
  timestamp: Date;
  knownItemIds: Set<string>;
}

/** ACL entry for items. */
interface AclEntry {
  type: string;
  value: string;
  accessType: string;
}

/** An external item payload. */
interface ExternalItem {
  acl: AclEntry[];
  properties: Record<string, unknown>;
  content: {
    value: string;
    type: "text" | "html";
  };
}

/**
 * Performs incremental sync between an external data source
 * and a Microsoft Graph external connection.
 */
class IncrementalSync {
  constructor(
    private readonly graphClient: Client,
    private readonly connectionId: string
  ) {}

  /**
   * Perform incremental sync by comparing source items against a checkpoint.
   * Only ingests new/modified items and deletes removed items.
   *
   * @param currentSourceItems - All items currently in the source system
   * @param lastCheckpoint - Checkpoint from the previous sync run
   */
  async runIncrementalSync(
    currentSourceItems: SourceItem[],
    lastCheckpoint: SyncCheckpoint
  ): Promise<void> {
    const sourceById = new Map(
      currentSourceItems.map((item) => [item.id, item])
    );

    // 1. Find new and modified items
    const toUpsert = currentSourceItems.filter(
      (item) =>
        !lastCheckpoint.knownItemIds.has(item.id) ||
        item.lastModified > lastCheckpoint.timestamp
    );

    // 2. Find deleted items (in previous checkpoint but not in current source)
    const toDelete = [...lastCheckpoint.knownItemIds].filter(
      (id) => !sourceById.has(id)
    );

    console.log(
      `Incremental sync: ${toUpsert.length} upserts, ${toDelete.length} deletes`
    );

    // 3. Upsert new/modified items
    for (const sourceItem of toUpsert) {
      const externalItem = mapToExternalItem(sourceItem);
      await this.graphClient
        .api(
          `/external/connections/${this.connectionId}/items/${sourceItem.id}`
        )
        .put(externalItem);
    }

    // 4. Delete removed items
    for (const itemId of toDelete) {
      try {
        await this.graphClient
          .api(
            `/external/connections/${this.connectionId}/items/${itemId}`
          )
          .delete();
      } catch (error: unknown) {
        const statusCode = (error as { statusCode?: number }).statusCode;
        if (statusCode !== 404) throw error;
        // Item already deleted — safe to ignore
      }
    }

    // 5. Update checkpoint
    const newCheckpoint: SyncCheckpoint = {
      timestamp: new Date(),
      knownItemIds: new Set(sourceById.keys()),
    };
    await this.saveCheckpoint(newCheckpoint);
  }

  /**
   * Hash-based change detection: only re-ingest items whose content hash changed.
   * More efficient than timestamp-based when source doesn't track modification dates.
   *
   * @param currentSourceItems - All items currently in the source system
   * @param previousHashes - Map of item ID → content hash from previous run
   * @returns Updated hash map to persist for the next run
   */
  async runHashBasedSync(
    currentSourceItems: SourceItem[],
    previousHashes: Map<string, string>
  ): Promise<Map<string, string>> {
    const updatedHashes = new Map(previousHashes);

    for (const sourceItem of currentSourceItems) {
      const currentHash = computeContentHash(sourceItem);

      if (previousHashes.get(sourceItem.id) === currentHash) {
        continue; // No changes — skip
      }

      const externalItem = mapToExternalItem(sourceItem);
      await this.graphClient
        .api(
          `/external/connections/${this.connectionId}/items/${sourceItem.id}`
        )
        .put(externalItem);

      updatedHashes.set(sourceItem.id, currentHash);
    }

    return updatedHashes;
  }

  /**
   * Persist a checkpoint for the next sync run.
   * Implement: save to database, blob storage, or local file.
   */
  private async saveCheckpoint(checkpoint: SyncCheckpoint): Promise<void> {
    console.log(
      `Checkpoint saved: ${checkpoint.timestamp.toISOString()}, ` +
        `${checkpoint.knownItemIds.size} items`
    );
    // Replace with your persistence mechanism (e.g., fs, database, blob)
  }
}

/** Map a source item to a Graph external item payload. */
function mapToExternalItem(source: SourceItem): ExternalItem {
  return {
    acl: [
      {
        type: "everyone",
        value: "everyone",
        accessType: "grant",
      },
    ],
    properties: {
      title: source.title,
      status: source.status,
      lastModifiedDate: source.lastModified.toISOString(),
      itemUrl: source.url,
      iconUrl: source.iconUrl ?? "",
    },
    content: {
      value: source.fullContent,
      type: "text",
    },
  };
}

/** Compute a SHA-256 hash of an item's key fields for change detection. */
function computeContentHash(item: SourceItem): string {
  const combined = [
    item.title,
    item.status,
    item.fullContent,
    item.lastModified.toISOString(),
  ].join("|");

  return createHash("sha256").update(combined, "utf-8").digest("hex");
}

export {
  IncrementalSync,
  computeContentHash,
  mapToExternalItem,
  SourceItem,
  SyncCheckpoint,
  ExternalItem,
};

File: copilot-connector/sample_codes/python/common-patterns/incremental-sync.py
# Incremental sync pattern: track changes and only re-ingest modified items
# Prerequisites: pip install msgraph-sdk azure-identity

import hashlib
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional

from msgraph import GraphServiceClient
from msgraph.generated.models.external_connectors.external_item import ExternalItem
from msgraph.generated.models.external_connectors.external_item_content import ExternalItemContent
from msgraph.generated.models.external_connectors.external_item_content_type import ExternalItemContentType
from msgraph.generated.models.external_connectors.acl import Acl
from msgraph.generated.models.external_connectors.acl_type import AclType
from msgraph.generated.models.external_connectors.access_type import AccessType
from msgraph.generated.models.external_connectors.properties import Properties
from msgraph.generated.models.o_data_errors.o_data_error import ODataError


@dataclass
class SourceItem:
    """Represents an item from the external data source."""

    id: str
    title: str
    status: str
    full_content: str
    url: str
    icon_url: Optional[str] = None
    last_modified: datetime = field(default_factory=lambda: datetime.now(timezone.utc))


@dataclass
class SyncCheckpoint:
    """Tracks sync state between incremental sync runs."""

    timestamp: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    known_item_ids: set[str] = field(default_factory=set)


class IncrementalSync:
    """Manages incremental sync for a Copilot Connector."""

    def __init__(self, graph_client: GraphServiceClient, connection_id: str) -> None:
        self._graph_client = graph_client
        self._connection_id = connection_id

    async def run_incremental_sync(
        self,
        current_source_items: list[SourceItem],
        last_checkpoint: SyncCheckpoint,
    ) -> SyncCheckpoint:
        """
        Perform incremental sync by comparing source items against a checkpoint.
        Only ingests new/modified items and deletes removed items.

        Returns:
            A new SyncCheckpoint reflecting the current state.
        """
        source_by_id = {item.id: item for item in current_source_items}
        previous_ids = last_checkpoint.known_item_ids

        # 1. Find new and modified items
        to_upsert = [
            item
            for item in current_source_items
            if item.id not in previous_ids
            or item.last_modified > last_checkpoint.timestamp
        ]

        # 2. Find deleted items (in previous checkpoint but not in current source)
        to_delete = [
            item_id for item_id in previous_ids if item_id not in source_by_id
        ]

        print(
            f"Incremental sync: {len(to_upsert)} upserts, {len(to_delete)} deletes"
        )

        # 3. Upsert new/modified items
        for source_item in to_upsert:
            external_item = _map_to_external_item(source_item)
            await self._graph_client.external.connections.by_external_connection_id(
                self._connection_id
            ).items.by_external_item_id(source_item.id).put(external_item)

        # 4. Delete removed items
        for item_id in to_delete:
            try:
                await self._graph_client.external.connections.by_external_connection_id(
                    self._connection_id
                ).items.by_external_item_id(item_id).delete()
            except ODataError as e:
                if e.response_status_code == 404:
                    pass  # Item already deleted — safe to ignore
                else:
                    raise

        # 5. Return updated checkpoint
        new_checkpoint = SyncCheckpoint(
            timestamp=datetime.now(timezone.utc),
            known_item_ids=set(source_by_id.keys()),
        )
        await self._save_checkpoint(new_checkpoint)
        return new_checkpoint

    async def run_hash_based_sync(
        self,
        current_source_items: list[SourceItem],
        previous_hashes: dict[str, str],
    ) -> dict[str, str]:
        """
        Hash-based change detection: only re-ingest items whose content hash changed.
        More efficient than timestamp-based when source doesn't track modification dates.

        Returns:
            Updated hash dictionary reflecting the current state.
        """
        updated_hashes = dict(previous_hashes)

        for source_item in current_source_items:
            current_hash = _compute_content_hash(source_item)

            if previous_hashes.get(source_item.id) == current_hash:
                continue  # No changes — skip

            external_item = _map_to_external_item(source_item)
            await self._graph_client.external.connections.by_external_connection_id(
                self._connection_id
            ).items.by_external_item_id(source_item.id).put(external_item)

            updated_hashes[source_item.id] = current_hash

        return updated_hashes

    async def _save_checkpoint(self, checkpoint: SyncCheckpoint) -> None:
        """Save checkpoint to persistent storage (database, blob storage, or file)."""
        # Implement: save to database, blob storage, or local file
        print(
            f"Checkpoint saved: {checkpoint.timestamp}, "
            f"{len(checkpoint.known_item_ids)} items"
        )


def _map_to_external_item(source: SourceItem) -> ExternalItem:
    """Convert a source item to a Microsoft Graph ExternalItem."""
    return ExternalItem(
        id=source.id,
        acl=[
            Acl(
                type=AclType.Everyone,
                value="everyone",
                access_type=AccessType.Grant,
            )
        ],
        properties=Properties(
            additional_data={
                "title": source.title,
                "status": source.status,
                "lastModifiedDate": source.last_modified.isoformat(),
                "itemUrl": source.url,
                "iconUrl": source.icon_url or "",
            }
        ),
        content=ExternalItemContent(
            value=source.full_content,
            type=ExternalItemContentType.Text,
        ),
    )


def _compute_content_hash(item: SourceItem) -> str:
    """Compute a SHA-256 hash of item content for change detection."""
    combined = (
        f"{item.title}|{item.status}|{item.full_content}|{item.last_modified.isoformat()}"
    )
    return hashlib.sha256(combined.encode("utf-8")).hexdigest()
